### ISSUE TITLE
CH58X(CH582M) BLE Support

### ISSUE THREAD

**Issue by Slappy2022** [ch32-rs/ch32-rs#12]
I've been playing around with BLE in 582m. 
Here's an rt using some elements from wch: https://github.com/Slappy2022/ch58x-ble-rt
I also got started on a hal: https://github.com/Slappy2022/ch58x-hal

Some notes...
I've only attempted to use peripherals, there are three examples in the rt with linux-based "centrals" to test. 
I've tried pin-based pairing as a one-off and it seems to work. 
Hand-crafted assembly is used for some safe mode registers. I'm not sure how/if the pac should handle this. 
There are a few things that belong in the pac that were absent or buggy, the hal hits some of these registers directly. 
I'm not sure how licenses should work here. The rt steals some things from the wch repo, I don't think this is proper. 
Critical section works in the default riscv runtime, but not in my ble runtime. 
ble doesn't work in the default riscv runtime, this seems to be related to the handling of "highcode", but I haven't figured it out. 

Please let me know how we might align this work with the rest of ch32-rs, assuming there's interest.

**Comment by andelf** [comment]
I need some time to look into it. I'll check it this weekend.

> There are a few things that belong in the pac that were absent or buggy

Yes, their SVD files are very buggy. So they are patched with svdtools.
The CH58x.svd is not revised yet.

> I'm not sure how licenses should work here. The rt steals some things from the wch repo, I don't think this is proper.

Since https://github.com/openwch/ch583/tree/main/EVT/EXAM/BLE/LIB is under Apache-2.0 license, make sure the libs/files you refer to are located in the openwch repo. Then it's OK to include them with a proper LICENSE.

> ble doesn't work in the default riscv runtime

This should be fixed by `pre_init` hook. If not feasible, we can invent a `-rt` crate for CH58x.

**Comment by andelf** [comment]
As with pac v0.2.0 release, most issues are fixed.

Check out https://github.com/ch32-rs/ch58x-hal, I've referred a lot from your hal crate. This is an `embassy`-style hal crate with embassy time driver support.

> Critical section works in the default riscv runtime, but not in my ble runtime.

Fixed in the ch58x-hal crate.

https://github.com/ch32-rs/ch58x-hal/blob/f3a25a2b6b66a26ec739197e00b81eaf7dd11aad/src/critical_section.rs#L6-L21

> This seems to be related to the handling of "highcode",

Added `highcode` support in the crate, using link script and `#[highcode]` macro.

Currently, I'm trying to port the BLE rt to it. At lease the RF_PHY example works.
